x-pack/plugins/beats/server/routes/api/register_enroll_beat_route.js

```diff
@@ -80,7 +81,7 @@ export function registerEnrollBeatRoute(server) {
       try {
         const enrollmentToken = request.headers['kbn-beats-enrollment-token'];
         const { token, expires_on: expiresOn } = await getEnrollmentToken(callWithInternalUser, enrollmentToken);
-        if (!token || token !== enrollmentToken) {
+        if (!token || !areTokensEqual(token, enrollmentToken)) {
```

Comment:
Does the `getEnrollmentToken()` return a token, even if one doesn't match the `enrollmentToken`?

Reply:
Sorry, this code might be confusing.

* `enrollmentToken` is what the user supplied to this API via the request header.
* `getEnrollmentToken()` returns the token stored in ES.

Then we compare the two.

Does that help?

Reply:
Oh, I see what you mean (after you edited the comment). `getEnrollmentToken` will return `undefined` for `token` if it doesn't find `enrollmentToken` in ES.

Reply:
Sorry about that, I wasn't being clear earlier. My concern is that if we don't find a token in ES, that we won't be executing the `areTokensEqual` function, so there will be a "timing difference", disclosing the existence of the key. In similar situations, I've done the comparison using a "blank" key to ensure that we're taking a static amount of time.

Reply:
We briefly discussed whether disclosing the mere existence of the key was an issue, and figured that wasn't a big deal in this case. If we want to not be susceptible to disclosing whether the key even exists, I think making sure the entirety of the call takes about e.g a multiple of 100 milliseconds. So if ES responds quickly with "no such key" in 5 ms, you'd wait 95ms, and if a network blip or having to route to another node causes it to take 101 ms, you'd wait another 99.

I tried finding a paper I've seen in the past re. how adding random noise isn't all that helpful. Couldn't re-find the paper, but this has a bit of coverage: https://blog.ircmaxell.com/2014/11/its-all-about-time.html

Reply:
Thanks @alexbrasetvik, I _think_ this sort of delay is what I ended up implementing in 7af9c2d. Can you take a look and let me know what you think? Thanks.

Reply:
(BTW, I'm trying to read that article but I keep getting jumped to the bottom of the page where there's a video).

Reply:
Okay, got the article into a PDF so I could read it. I read the note on random delays. I understand now that what I implemented in 7af9c2d is not effective because it just adds noise that can be averaged out. I'm looking into making the delay dependent on the `enrollmentToken` itself.